sktime/annotation/hmm.py

```diff
@@ -0,0 +1,275 @@
+            Thus each row should sum to 1.
+        num_obs : (int) the number of observations (m)
+        num_states : (int) the number of hidden states (n)
+
+        Returns
+        -------
+        trans_prob : (np.ndarray) an nxm dimensional array which represents the
+            maximum probability of the hidden state of observation m is state n.
+        trans_id : (np.ndarray) a nxm dimensional array which for each observation
+            "i" and state "j" the i,j entry records the state_id of the most
+            likely state that could have led to the hidden state being "i" for
+            observation "j".
+        """
+        # trans_prob represents the maximum probability of being in that
+        # state at that stage
+        trans_prob = np.zeros((num_states, num_obs))
+        trans_prob[:, 0] = np.log(initial_probs)
+
+        # trans_id is the index of the state that would have been the most
+        # likely preceeding state.
+        trans_id = np.zeros((num_states, num_obs), dtype=np.int32)
+
+        # use Vertibi Algorithm to fill in trans_prob and trans_id:
+        for i in range(1, num_obs):
+            # use log probabilities to try to keep nums reasonable -Inf
+            # means 0 probability
+            paths = np.zeros((num_states, num_states))
+            for j in range(num_states):
+                paths[j, :] += trans_prob[:, i - 1]  # adds prev trans_prob column-wise
+                paths[:, j] += np.log(emi_probs[:, i])  # adds log(probs_sub) row-wise
+            paths += np.log(
+                transition_prob_mat
+            )  # adds log(transition_prob_mat) element-wise
+            trans_id[:, i] = np.argmax(paths, axis=0)
+            trans_prob[:, i] = np.max(paths, axis=0)
+
+        if np.any(np.isinf(trans_prob[:, -1])):
+            raise ValueError("Change parameters, the distribution doesn't work")
+
+        return trans_prob, trans_id
+
+    @classmethod
+    def _make_emission_probs(
+        cls, emission_funcs: list, observations: np.ndarray
+    ) -> np.ndarray:
+        """Calculate the prob each obs comes from each hidden state."""
+        # assign emission probabilities from each state to each position:
+
+        emi_probs = np.zeros(shape=(len(emission_funcs), len(observations)))
+        for state_id, emission_tuple in enumerate(emission_funcs):
+            emission_func = emission_tuple[0]
+            kwargs = emission_tuple[1]
+            emi_probs[state_id, :] = np.array(
+                [emission_func(x, **kwargs) for x in observations]
+            )
+        return emi_probs
+
+    def _hmm_viterbi_label(self) -> np.array:
+        """Assign hidden state ids to all observations based on most likely path.
+
+        Parameters
+        ----------
+        self - an HMM instance which already has already
+
+        Returns
+        -------
+            -hmm_fit: a 2xn array with the first column being position and the second
```

Comment:
typo?

Reply:
Yes, also implemented Viterbi in another situation, and the comment was more relevant there. (biological annotation)